--- a/Cargo.toml
+++ b/Cargo.toml
@@ -17,7 +17,7 @@ default-run = "hnefatafl-server-full"
 [dev-dependencies]
 criterion = { version = "0.7", features = ["html_reports"] }
 futures = "0.3"
-iced = { git = "https://github.com/iced-rs/iced", features = ["smol"] }
+iced = { git = "https://github.com/iced-rs/iced", features = ["smol", "fira-sans"] }
 rodio = "0.21"
 webbrowser = { version = "1", features = ["hardened"] }
 
