--- a/cmake/external/helper_functions.cmake
+++ b/cmake/external/helper_functions.cmake
@@ -22,6 +22,7 @@ macro(onnxruntime_fetchcontent_makeavailable)
   )
   foreach(__cmake_contentName IN ITEMS ${ARGV})
     string(TOLOWER ${__cmake_contentName} __cmake_contentNameLower)
+    message("checking for ${__cmake_contentName}")
 
     # If user specified FETCHCONTENT_SOURCE_DIR_... for this dependency, that
     # overrides everything else and we shouldn't try to use find_package() or
@@ -134,6 +135,9 @@ macro(onnxruntime_fetchcontent_makeavailable)
       unset(__cmake_haveFpArgs)
     endif()
 
+
+    message("checking for ${__cmake_contentName} step 2")
+
     FetchContent_GetProperties(${__cmake_contentName})
     if(NOT ${__cmake_contentNameLower}_POPULATED)
       FetchContent_Populate(${__cmake_contentName})
@@ -184,6 +188,9 @@ macro(onnxruntime_fetchcontent_makeavailable)
           set(CMAKE_SKIP_INSTALL_RULES FALSE)
       endif()
 
+
+      message("finished checking for ${__cmake_contentName}")
+
       unset(__cmake_srcdir)
       unset(__cmake_contentDetails)
       unset(__cmake_arg_SOURCE_SUBDIR)
